src/sagas/dependency.saga.js

```diff
@@ -0,0 +1,325 @@
+    }
+  }
+}
+
+function* updateDependency({ projectId, dependencyName, latestVersion }) {
+  console.log('lock:', lock);
+  console.log('queue:', queue);
+  if (!(yield call(requestPackageJsonLockForProjectId, projectId))) {
+    yield call(enqueue, {
+      action: 'install',
+      projectId,
+      dependencyName,
+      version: latestVersion,
+      updating: true,
+    });
+    yield cancel();
+  } else {
+    const notificationId = `${projectId}-update-${dependencyName}`;
+    yield put(
+      showNotification(notificationId, {
+        title: 'Updating dependency...',
+        message: `${dependencyName}@${latestVersion}`,
+      })
+    );
+
+    const projectPath = yield select(getPathForProjectId, projectId);
+    try {
+      yield call(installDependency, projectPath, dependencyName, latestVersion);
+      yield put(completeNotification(notificationId));
+      yield put(
+        updateDependencyFinish(projectId, dependencyName, latestVersion)
+      );
+    } catch (err) {
+      yield call([console, 'error'], 'Failed to update dependency', err);
+      yield put(failNotification(notificationId, err));
+      yield put(updateDependencyError(projectId, dependencyName));
+    }
+  }
+}
+
+function* deleteDependency({ projectId, dependencyName }) {
+  console.log('lock:', lock);
+  console.log('queue:', queue);
+  if (!(yield call(requestPackageJsonLockForProjectId, projectId))) {
+    yield call(enqueue, { action: 'uninstall', projectId, dependencyName });
+    yield cancel();
+  } else {
+    const notificationId = `${projectId}-delete-${dependencyName}`;
+    yield put(
+      showNotification(notificationId, {
+        title: 'Deleting dependency...',
+        message: dependencyName,
+      })
+    );
+
+    const projectPath = yield select(getPathForProjectId, projectId);
+    try {
+      yield call(uninstallDependency, projectPath, dependencyName);
+      yield put(completeNotification(notificationId));
+      yield put(deleteDependencyFinish(projectId, dependencyName));
+    } catch (err) {
+      yield call([console, 'error'], 'Failed to delete dependency', err);
+      yield put(failNotification(notificationId, err));
+      yield put(deleteDependencyError(projectId, dependencyName));
+    }
+  }
+}
+
+function* addDependencies({ projectId, notificationId, dependencies }) {
+  yield put(
+    updateNotification(notificationId, { title: 'Installing dependencies...' })
+  );
+
+  const projectPath = yield select(getPathForProjectId, projectId);
+  try {
+    yield call(installDependencies, projectPath, dependencies);
+    const installedDependencies = yield call(
+      loadProjectDependencies,
+      projectPath,
+      dependencies
+    );
+    yield put(completeNotification(notificationId));
+    yield put(addDependenciesFinish(projectId, installedDependencies));
+  } catch (err) {
+    yield put(failNotification(notificationId, err));
+    yield put(addDependenciesError(projectId, dependencies));
```

Comment:
Hopefully my thoughts about removing notification actions makes sense :) in this case, it's clear that the adjacent actions could trigger those notifications, so we'd just need the dependency-focused ones :)

I know when I started using Redux, it felt wrong to me to have a single action affect multiple parts of the state; I was genuinely surprised when I learned that the best practice was to consolidate actions. But, I've been following that convention for a couple years now, and it's **so much nicer**.

Reply:
I'm excited to come to that same level of appreciation 😄 I'm still pretty new to the Redux way of thinking but it's already been as exciting of a paradigm shift as functional programming (currying and composing and all that other good stuff).

Reply:
Yeah, I really like Redux. I have a pet theory that the reason many folks don't enjoy working with Redux is because they're trying to use existing "best-practices" from MVC frameworks with it, and Redux is kind of its own thing, with its own best practices.

I've certainly worked on codebases where Redux was this high-friction PITA, but I have a totally different experience with it on my side-projects when I can set actions up to mirror user actions (and the resulting asynchronous changes)